core/string_builder.cpp

```diff
@@ -0,0 +1,97 @@
+/* Copyright (c) 2007-2017 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2017 Godot Engine contributors (cf. AUTHORS.md)    */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+#include "string_builder.h"
+
+#include <string.h>
+
+StringBuilder &StringBuilder::append(const String &p_string) {
+
+	strings.push_back(p_string);
+	appended_strings.push_back(-1);
+
+	string_length += p_string.length();
+
+	return *this;
+}
+
+StringBuilder &StringBuilder::append(const char *p_cstring) {
+
+	int32_t len = strlen(p_cstring);
+
+	c_strings.push_back(p_cstring);
+	appended_strings.push_back(len);
+
+	string_length += len;
+
+	return *this;
+}
+
+String StringBuilder::as_string() const {
+
+	CharType *buffer = memnew_arr(CharType, string_length);
+
+	int current_position = 0;
+
+	int godot_string_elem = 0;
+	int c_string_elem = 0;
+
+	for (int i = 0; i < appended_strings.size(); i++) {
+		if (appended_strings[i] == -1) {
+			// Godot string
+			const String &s = strings[godot_string_elem];
+
+			memcpy(buffer + current_position, s.ptr(), s.length() * sizeof(CharType));
+
+			current_position += s.length();
+
+			godot_string_elem++;
+		} else {
+			// This is a bit ugly, because c strings are char * and
+			// CharType is actually wchar_t we need to convert it somehow.
+			// We will use Godot's String for that...
+
+			const char *s = c_strings[c_string_elem];
+
+			String converted_string = String::utf8(s, appended_strings[i]);
```

Comment:
You should not assume that this is utf8. Godot simply does not have this datatype around in this format, as it's always converted when parsed. I would assume it a regular ascii C string. and copy char -> wchar_t with a for() loop. It will be much faster for existing use cases we intend to use this for.